### PR TITLE
Add State management MVI Post

### PR DESCRIPTION
**Pull request by riggaroo**
Tested locally and changed image bases etc. 

Screenshot 2021-11-06 at 19 38 57

### CODE REVIEW

File: _posts/2021-11-08-android-state-management-mvi.md [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+    photo: /assets/images/rebecca_franks.jpg
+  - name: Kamil Slesinski
+    title: Software Development Engineer IV
+    url: https://www.linkedin.com/in/kamilslesinski/
+    photo: /assets/images/kamil_slesinski.jpeg
+---
+
+## History ⌛️
+
+On the GoDaddy Studio Android app, we started out with a fresh codebase around the end of 2018. Right when Kotlin was starting to become the popular choice and the community had moved on from choosing MVP as the preferred architecture choice, to MVVM. We decided to adopt MVVM where it made sense, but we were intrigued by the prospects of MVI/[Unidirectional data flow](https://en.wikipedia.org/wiki/Unidirectional_Data_Flow_(computer_science)). After watching a few presentations on the topic, we proceeded to design our Canvas Editor screen using a homegrown version of MVI, very similar to the [code](https://github.com/oldergod/android-architecture) described in this [presentation](https://www.youtube.com/watch?v=64rQ9GKphTg&ab_channel=oldergod).
```
Comment (kawikabader) [inline _posts/2021-11-08-android-state-management-mvi.md:23]:
```suggestion
The GoDaddy Studio Android app began with a fresh codebase around the end of 2018. This was around the same time Kotlin was starting to become a popular choice and the community had moved on from choosing MVP as the preferred architecture choice, to MVVM. We decided to adopt MVVM where it made sense, but we were intrigued by the prospects of MVI/[Unidirectional data flow](https://en.wikipedia.org/wiki/Unidirectional_Data_Flow_(computer_science)). After watching a few presentations on the topic, we proceeded to design our Canvas Editor screen using a homegrown version of MVI, very similar to the [code](https://github.com/oldergod/android-architecture) described in this [presentation](https://www.youtube.com/watch?v=64rQ9GKphTg&ab_channel=oldergod).
```

**Review by kawikabader — CHANGES_REQUESTED** [review]
Would it make sense to have a small section near the top of this post explaining what MVP, MVVM, and MVI are?

**Review by kawikabader — APPROVED** [review]

File: _posts/2021-11-08-android-state-management-mvi.md [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+# MVI / Unidirectional Data Flow
+
+Now that we understand where potential issues can come from with MVVM and multiple LiveData observables, let’s talk a bit about Unidirectional Data Flow or MVI (Model-View-Intent) and our experience with it.
+
+With Unidirectional Data Flow everything is modelled around a single state object. This state is immutable, to prevent unwanted side effects. There is a clear flow of how data changes through the system, since it only happens in a single direction.
+
+
+{:refdef: style="text-align: center;"}
+<figure>
+  <img src="{{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="my alt text"/>
```
Comment (tarkatronic) [inline _posts/2021-11-08-android-state-management-mvi.md:131]:
```suggestion
  <img src="{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="my alt text"/>
```
Jekyll was throwing a Liquid syntax error on this. Also, any chance this alt text can be updated?

**Review by tarkatronic — COMMENTED** [review]
Just one minor syntax fix; otherwise, let's go! 🚢 :shipit:

File: _posts/2021-11-08-android-state-management-mvi.md (hunk 2 above)
Comment (riggaroo) [reply _posts/2021-11-08-android-state-management-mvi.md:131]:
```suggestion
  <img src="{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="Diagram showing Unidirectional Data Flow"/>
```

File: _posts/2021-11-08-android-state-management-mvi.md [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+- What should happen when the project is loading up?
+- Has an error occurred?
+- What is the currently selected tool on screen?
+- What are the layers and properties that should be displayed in this project?
+
+Not only is there the current state of what should be rendered on screen, but there are many different interactions that need to mutate this state - scaling gestures, button taps, colour changes etc. 
+
+## MVVM and the issues we faced 😣
+
+A couple of years ago, the Android community began advocating for using **some kind of architecture,** as we all realized that placing all your logic inside the Activity, was a perfect recipe for the messiest spaghetti bolognaise you could imagine (delicious but not very elegant).
```
[the remark this answers is not in the document]
Comment (riggaroo) [reply _posts/2021-11-08-android-state-management-mvi.md:46]: Apparently they are both right 😬  https://www.newgrounds.com/bbs/topic/1043082 haha

File: _posts/2021-11-08-android-state-management-mvi.md [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+
+First came **MVP** (Model-View-Presenter) and there were great benefits to applying this architecture to our apps: we got testability, separation of concerns and the ability to re-use logic on other screens.
```
[the remark this answers is not in the document]
Comment (riggaroo) [reply _posts/2021-11-08-android-state-management-mvi.md:48]: This looks like both are correct - https://www.grammarly.com/blog/comma-before-and/ just need to be consistent. Will stick without the ,

File: _posts/2021-11-08-android-state-management-mvi.md [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+- We now have a standardized ViewModel implementation across the app
+
+With every new framework or architecture decision, there is likely never going to be the solution that fits everyone, and adopting something like the Mobius Framework also comes with these disadvantages to adopting:
+
+- It requires quite a bit of boilerplate code to set up. We’ve solved this using shared code templates to generate most of the classes we need. 
+- New framework for people to learn. Any new engineer needs to spend a bit of time learning about the framework, working through some examples and implementing a feature with it. 
+
+# Summary 🎉
+
+We’ve learnt a lot over the past couple of years, and I don’t doubt that we won’t learn more in the future.
```
[the remark this answers is not in the document]
Comment (riggaroo) [reply _posts/2021-11-08-android-state-management-mvi.md:411]: https://www.grammarly.com/blog/learned-learnt/ both are right again 😄  Leaving as learnt.

File: _posts/2021-11-08-android-state-management-mvi.md [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+- There is no single snapshot to be able to recreate the UI from easily.
+- This also then begs the question, are we handling all the potential cases here?
+
+One way to potentially improve on this behaviour, is to use data classes that contain the state information in one class, and expose only a singular LiveData object to the UI. This would help solve the issue of having multiple observables emitting different state that we’d need to keep track of, and its a building block of how MVI can work, as we will explore next. 
+
+# MVI / Unidirectional Data Flow
+
+Now that we understand where potential issues can come from with MVVM and multiple LiveData observables, let’s talk a bit about Unidirectional Data Flow or MVI (Model-View-Intent) and our experience with it.
+
+With Unidirectional Data Flow everything is modelled around a single state object. This state is immutable, to prevent unwanted side effects. There is a clear flow of how data changes through the system, since it only happens in a single direction.
```
[the remark this answers is not in the document]
Comment (riggaroo) [reply _posts/2021-11-08-android-state-management-mvi.md:126]: Another US vs British english thing apparently - https://www.grammarly.com/blog/modeling-or-modelling/

File: _posts/2021-11-08-android-state-management-mvi.md [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+
+As time went on, we started facing issues with the MVI approach we used. The way we implemented it, unfortunately allowed for race conditions and caused a lot of headaches for the team. It was at this point that we decided to investigate using existing frameworks for unidirectional data flow, and see if they could solve the problems we were seeing with ours. After evaluating a few, we settled on using the [Spotify Mobius Framework](https://github.com/spotify/mobius) for various reasons which we will dive into in this blog post.
+
+## State Management 
+
+Let’s step back a bit and learn about the issues we faced with our approaches over the past couple of years. What does it mean when we talk about State Management for UI? State management refers to keeping track of how a user interface should look and how it should react to different user inputs. On a complex screen with plenty of buttons, gestures and text input - managing the state is a full time concern.
```
[the remark this answers is not in the document]
Comment (riggaroo) [reply _posts/2021-11-08-android-state-management-mvi.md:29]:
```suggestion
Let’s step back a bit and learn about the issues we faced with our approaches over the past couple of years. What does it mean when we talk about State Management for UI? State management refers to keeping track of how a user interface should look and how it should react to different user inputs. On a complex screen with plenty of buttons, gestures and text input - managing the state is a full-time concern. 
```

**Review by tarkatronic — APPROVED** [review]

**Pull request merged**